refactor: replace axios with native fetch for binary download

harshjainh2j opened the pull request:
refactor: replace axios with undici for binary download

Removed axios as a dependency and replaced it with undici's fetch API to download and extract the binary file. This reduces package bloat and improves compatibility with Node.js versions below 18.

- Uses undici's fetch for HTTP requests
- Streams the response directly into tar extractor
- Preserves existing error handling and logging
- Tested to ensure proper binary installation

taoeffect commented on scripts/binary-install.js:
```diff
@@ -102,4 +108,4 @@ class Binary {
 }
 
-module.exports.Binary = Binary
+module.exports.Binary = Binary
```
Ensure newline at EOF

corrideat replied:
Missing end-of-line

taoeffect commented on scripts/binary-install.js:
```diff
@@ -7,8 +7,9 @@ const process = require('node:process')
 const { join, resolve } = require('path')
 const { spawnSync } = require('child_process')
+const { Readable } = require('node:stream') // Required for stream conversion
+const { fetch } = require('undici') // Keep undici for fetch
```
Use built-in `fetch`

corrideat replied:
`fetch` should be defined without using `undici`, so this line and the new `undici` can be removed.

corrideat commented on bin/chel:
This shouldn't be deleted

taoeffect commented on package.json:
```diff
@@ -20,9 +20,9 @@
-    "axios": "0.27.2",
     "rimraf": "3.0.2",
     "tar": "6.1.11",
+    "undici": "7.12.0",
```
Remove this dependency

corrideat replied:
As mentioned, this can be removed.

corrideat commented on bin/chel.exe:
I guess it's fine to add this file, but it should just be a placeholder.

taoeffect reviewed (changes requested):
Thanks for the PR @harshjainh2j! I left some feedback.

Also, sorry about this, but please avoid running `npm install`, as there's an open issue #37 that causes a local run of `npm install` to delete `bin/chel` and replace it with a binary - so those changes here need to be reverted.

corrideat reviewed (changes requested):
Thanks a lot for your help @harshjainh2j! The PR looks fine, but please note my minor comments.

corrideat commented:
> Also, sorry about this, but please avoid running npm install, as there's an open issue https://github.com/okTurtles/chel/issues/37 that causes a local run of npm install to delete bin/chel and replace it with a binary - so those changes here need to be reverted.'

I'm afraid `npm install` is needed in this case since `package-lock.json` is modified, but maybe try running `npm install --ignore-scripts`, or manually verify the contents of `bin`.

corrideat commented on scripts/binary-install.js:
```diff
@@ -64,23 +65,28 @@ class Binary {
+      }
+
+      // Convert web stream to Node.js stream (CRITICAL FIX)
+      const nodeStream = Readable.fromWeb(response.body)
```
This line in itself is fine, but I've seen a few Node.js bugs related to using this specific API (I mean that the Node.js logic was, at some point at least, buggy, for converting from web streams to native streams).

harshjainh2j commented:
Hey ! @corrideat , @taoeffect , I have removed undici and used the global fetch instead . and ensured no accidental changes to bin/chel after npm install, and manually verified its contents.

taoeffect commented:
Thanks @harshjainh2j! I think you'll need to manually use some `git` commands to prevent these changes from showing up though, as `chel` is still being modified, and `chel.exe` is being added (keep in mind this will happen each time you run `npm install` until #37 is closed, so don't run that command):

![Screenshot 2025-07-30 at 9 55 29 AM](https://github.com/user-attachments/assets/20c45f4f-4c21-45dc-9456-2574737c5914)

harshjainh2j commented:
Hey!  @corrideat @taoeffect ,Thanks for the feedback  , I’ve made all the requested changes:

1. Added a line break at the end of scripts/binary-install.js

2. Restored bin/chel from master

3. Replaced bin/chel.exe with the same placeholder as bin/chel

taoeffect reviewed (approved):
Alright! Nicely done @harshjainh2j! 

Testing this will require trying out an update I suppose, so it's unclear whether this will work, but it looks like it should so I'll approve and merge.

Thanks for the PR! 😄

corrideat commented:
> Testing this will require trying out an update

Not really. Running `npm install` should be a sufficient test, or, alternatively, `npm pack`, and then `npm install <path-to-package>` on a different project.
